src/annot/spec.ml

```diff
@@ -0,0 +1,98 @@
+open Types
+open Fmt
+
+(*
+  text:
+    (local index)
+    (global index)
+    (binder index)
+    string_id
+
+  binary:
+    (local index)
+    (global index)
+    (binder index)
+*)
+
+type nonrec binpred =
+  | Ge
+  | Gt
+  | Le
+  | Lt
+  | Eq
+  | Neq
+
+type nonrec unconnect = Neg
+
+type nonrec binconnect =
+  | And
+  | Or
+  | Imply
+  | Equiv
+
+type nonrec binder =
+  | Forall
+  | Exists
+
+type nonrec binder_type = num_type
+
+type 'a prop =
+  | Const : bool -> 'a prop
+  | BinPred : binpred * 'a term * 'a term -> 'a prop
+  | UnConnect : unconnect * 'a prop -> 'a prop
+  | BinConnect : binconnect * 'a prop * 'a prop -> 'a prop
+  | Binder : binder * binder_type * string option * 'a prop -> 'a prop
+
+and 'a term =
+  | Int32 : int32 -> 'a term
+  | Var : text indice -> text term
+  | GlobalVar : 'a indice -> 'a term
+  | BinderVar : 'a indice -> 'a term
+  | Result : 'a term
```

Comment:
Why are you making `prop` and `term` mutually recursive ? You could define `term` first and then `prop`, right?

Reply:
I'm considering using props in if-else expression. But that's incremental and I'm just leaving the place of that.

For the GADT, it's for writing the production `| Var : text indice -> text term`. I think it's good for two reasons:
- It can allow using string ids in text annotations, which may facilitate writing specifications.
- And we don't need to define separate types for text contract and binary contract. Text contract can be of type `text contract` (which additionally allows the constructor `Var`) and binary can be `binary contract`.

(But that also means we have to parse contracts in text format, and transform it to binary format through the group+assign+rewrite process. What do you think of this?)

Reply:
> I'm considering using props in if-else expression. But that's incremental and I'm just leaving the place of that.

OK fine with me. I'd still remove the mutual recursion for now to make everything simpler as a start but you can keep it if you want.

Reply:
I think I can make a "clean" version once I have the whole process running :) I admit these commits are mainly for visualization rather than works finished.

Reply:
> Yes, that's the plan, we'll have to do this at some point.

In our earlier discussion we said to parse the contract in binary format. I guess it's a slip of the tongue?

Reply:
We can start by parsing it it the binary format but it'll produce `binary prop`/`binary term` directly. But it'll be hard to test because I'm not sure you'll want to write binary propositions by hand. :-)

So indeed, it's probably better to start by parsing them in the text format, producing `text prop`/`text term`. Then do the translation (you can also start by not doing the translation of names and only translate everything directly with an identity function that fails on text identifiers). Then we'll do the generation.

I'm not sure if people are likely to write annotations by hand in the binary format so it probably doesn't matter in the beginning. It may be useful if we want to try to compile E-ACSL annotations to Wasm annotations, but's that's probably a full PhD work haha.